### DISCUSSION TITLE
How to inject custom data to pug

### DISCUSSION THREAD

**Discussion by Mizan-Rifat** [webdiscus/pug-plugin#15] [Q&A]
Is there any way to inject custom data in pug files like htmlWebpackPlugin.options.data in htmlWebpackPlugin??

**Comment by webdiscus** [comment]
Hello @Mizan-Rifat 

If you use the `htmlWebpackPlugin`  then is enough  to use the [@webdiscus/pug-loader](https://github.com/webdiscus/pug-loader) only.
The `PugPlugin.loader` contains same `@webdiscus/pug-loader`.

You can see the [passing data in HtmlWebpackPlugin](https://github.com/webdiscus/pug-loader#passing-data-in-htmlwebpackplugin) by using `@webdiscus/pug-loader`.

webpack.config.js
```js
module.exports = {
  plugins: [
    new HtmlWebpackPlugin({
      title: 'The some page', // avaliable in pug as `htmlWebpackPlugin.options.title`
      data: {
        key: 'value', //  avaliable in pug as `htmlWebpackPlugin.options.data.key`
      },
      template: path.join(__dirname, 'src/index.pug'), 
      filename: 'index.html',
    }),
  ],
  module: {
    rules: [
      {
        test: /\.pug$/,
        loader: '@webdiscus/pug-loader',
        options: {
           method: 'render', // use this method to render into static HTML
        }
      },
    ],
  },
};
```

src/index.pug
```pug
- const externalOptions = htmlWebpackPlugin.options;
- const externalData = htmlWebpackPlugin.options.data;

div Custom title is #{externalOptions.title}
div The value of 'key' is #{externalData.key}
```

### Passing data in pug using pug-plugin

You can pass the data into single pug via query parameters:

```js
const PugPlugin = require('pug-plugin');
module.exports = {
  entry: {
    'page01': 'page01.pug?myData=' + JSON.stringify({ title: 'About', options: { uuid: 'abc123' } }), 
  },
  plugins: [
    new PugPlugin(),
  ],
  module: {
    rules: [
      {
        test: /\.pug$/,
        loader: PugPlugin.loader,
        options: {
           method: 'render',
        }
      },
    ],
  },
};
```
page01.pug
```pug
div Custom title is #{myData.title}
div The UUID is #{myData.options.uuid}
```
This is absolutely legal and [documented feature of webpack](https://webpack.js.org/api/loaders/#thisresourcequery).

For more details, see [How to pass custom data from webpack.config.js into the pug template](https://github.com/webdiscus/pug-loader#passing-data-into-template).

**Comment by webdiscus** [comment]
@Mizan-Rifat I answered your question, did you succeed?

**Comment by Mizan-Rifat** [comment]
@webdiscus yes. Thanks for the reply...

**Comment by Mizan-Rifat** [comment]
@webdiscus hi, in htmlwebpackplugin i can pass a function in data. But in pug-plugin i can't pass the function. Is there any way to do it??

**Comment by webdiscus** [comment]
Hello @Mizan-Rifat,

in the next version will be added supports a function in pug-loader option `data`.
The option `data` is global for all templates.

For example,
webpack.config.js module.rules
```js
{
  loader: PugPlugin.loader,
  options: {
    method: 'render',
    data: {
      myData: {
        someVar: 'some text',
        sum(arg1, arg2) {
          return arg1 + arg2;
        },
      }
    },
  },
},
```
Usage the passed data In pug template:
```pug
p The external variable #{myData.someVar}
p The sum is: #{myData.sum(10, 20)}
```

Would this be the solution for you?

**Comment by Mizan-Rifat** [comment]
@webdiscus  yes. that would be great..

**Comment by webdiscus** [comment]
Hello @Mizan-Rifat,

new version 2.0.0 supports a function in `data` option of `pug-loader`, see the [test case](https://github.com/webdiscus/pug-loader/tree/master/test/cases/pass-data-method-render)

**Note**: you can require a JS module  and call a lib function directly in pug, see the [test case](https://github.com/webdiscus/pug-plugin/blob/master/test/cases/require-scripts-render/src/pages/index.pug)
Perhaps, this feature can solve your problem for  the single pug template.

> Limitation by the compile method. 
> A string representing the source code of the function is limited by the function.toString(), see [examples](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString#examples). 
> For native work of the function passed via the data loader option, use the `render` method.

### The bonus
The new version supports the require() of styles and scripts directly in pug without having to define them in a webpack entry.

```pug
html
  head
    link(rel='stylesheet' href=require('./styles.scss'))
    script(src=require('./main.js'))
  body
```

Output:
```html
<!DOCTYPE html>
<html>
  <head>
    <link rel="stylesheet" href="/assets/css/styles.05e4dd86.css">
    <script src="/assets/js/main.f4b855d8.js"></script>
  </head>
  <body></body>
</html>
```

webpack.config.js
```js
const PugPlugin = require('pug-plugin');
module.exports = {
  output: {
    publicPath: '/',
    // js output filename 
    filename: 'assets/js/[name].[contenthash:8].js',
  },
  entry: {
    // all scripts and styles can be defined directly in pug
    'index': './src/index.pug',
  },
  plugins: [
    new PugPlugin({
      modules: [
        PugPlugin.extractCss({
          // css output filename 
          filename: 'assets/css/[name].[contenthash:8].css',
        }),
      ],
    }),
  ],
  module: {
    rules: [
      {
        test: /\.(pug)$/,
        loader: PugPlugin.loader,
        options: {
          method: 'render',
        }
      },
      {
        test: /\.scss$/i,
        use: ['css-loader', 'scss-loader'],
      },
    ],
  },
};
```
